src/reformat.c

```diff
@@ -443,6 +457,67 @@ avifResult avifImageRGBToYUV(avifImage * image, const avifRGBImage * rgb)
                 }
             }
         }
+    } else if (!converted && isGray) {
+        const uint32_t grayPixelBytes = state.rgb.pixelBytes;
+        const uint32_t offsetBytesGray = state.rgb.offsetBytesGray;
+        const uint32_t offsetBytesA = state.rgb.offsetBytesA;
+        const uint32_t grayRowBytes = rgb->rowBytes;
+        const float grayMaxChannelF = state.rgb.maxChannelF;
+        uint8_t * yPlane = image->yuvPlanes[AVIF_CHAN_Y];
+        const uint32_t yRowBytes = image->yuvRowBytes[AVIF_CHAN_Y];
+        for (uint32_t j = 0; j < image->height; ++j) {
+            for (uint32_t i = 0; i < image->width; ++i) {
+                float g;
+                if (state.rgb.channelBytes > 1) {
+                    g = *(uint16_t *)&rgb->pixels[offsetBytesGray + i * grayPixelBytes + (j * grayRowBytes)];
+                } else {
+                    g = rgb->pixels[offsetBytesGray + i * grayPixelBytes + (j * grayRowBytes)];
+                }
+                if (alphaMode != AVIF_ALPHA_MULTIPLY_MODE_NO_OP) {
+                    float a;
+                    if (state.rgb.channelBytes > 1) {
+                        a = *((uint16_t *)(&rgb->pixels[offsetBytesA + (i * grayPixelBytes) + (j * grayRowBytes)])) / grayMaxChannelF;
+                    } else {
+                        a = rgb->pixels[offsetBytesA + (i * grayPixelBytes) + (j * grayRowBytes)] / grayMaxChannelF;
+                    }
+
+                    if (alphaMode == AVIF_ALPHA_MULTIPLY_MODE_MULTIPLY) {
+                        if (a == 0) {
+                            g = 0;
+                        } else if (a < 1.0f) {
+                            g *= a;
+                        }
+                    } else {
+                        // alphaMode == AVIF_ALPHA_MULTIPLY_MODE_UNMULTIPLY
+                        if (a == 0) {
+                            g = 0;
+                        } else if (a < 1.0f) {
+                            g /= a;
+                            g = AVIF_MIN(g, 1.0f);
+                        }
+                    }
+                }
+                g = avifRoundf(AVIF_CLAMP(g, 0.0f, state.yuv.maxChannel));
+                if (state.yuv.channelBytes > 1) {
+                    uint16_t * pY = (uint16_t *)&yPlane[(i * 2) + j * yRowBytes];
+                    *pY = (uint16_t)g;
+                } else {
+                    yPlane[i + (j * yRowBytes)] = (uint8_t)g;
+                }
+            }
+        }
+        // Set the chroma planes to 128 if any.
```

Comment:
128 is only correct for 8 bits. For high bit depths (10, 12, and 16) this should be 2^(depth - 1). And we will need a memset16 function.